packages/template-retail-react-app/app/pages/checkout/partials/payment.jsx

```diff
@@ -4,67 +4,140 @@
 import AddressDisplay from '../../../components/address-display'
 import {PromoCode, usePromoCode} from '../../../components/promo-code'
+import {API_ERROR_MESSAGE} from '../../../constants'
 
 const Payment = () => {
     const {formatMessage} = useIntl()
+    const {data: basket} = useCurrentBasket()
+    const selectedShippingAddress = basket?.shipments && basket?.shipments[0]?.shippingAddress
+    const selectedBillingAddress = basket?.billingAddress
+    const appliedPayment = basket?.paymentInstruments && basket?.paymentInstruments[0]
+    const [billingSameAsShipping, setBillingSameAsShipping] = useState(
+        isMatchingAddress(selectedBillingAddress, selectedShippingAddress)
+    )
+    const {mutateAsync: addPaymentInstrumentToBasket} = useShopperBasketsMutation(
+        'addPaymentInstrumentToBasket'
+    )
+    const {mutateAsync: updateBillingAddressForBasket} = useShopperBasketsMutation(
+        'updateBillingAddressForBasket'
+    )
+    const {mutateAsync: removePaymentInstrumentFromBasket} = useShopperBasketsMutation(
+        'removePaymentInstrumentFromBasket'
+    )
+    const showToast = useToast()
+    const showError = () => {
+        showToast({
+            title: formatMessage(API_ERROR_MESSAGE),
+            status: 'error'
+        })
+    }
+
+    const {step, STEPS, goToStep, goToNextStep} = useCheckout()
 
-    const {
-        step,
-        checkoutSteps,
-        setCheckoutStep,
-        selectedShippingAddress,
-        selectedBillingAddress,
-        selectedPayment,
-        getPaymentMethods,
-        removePayment
-    } = useCheckout()
-
-    const {
-        paymentMethodForm,
-        billingAddressForm,
-        billingSameAsShipping,
-        setBillingSameAsShipping,
-        reviewOrder
-    } = usePaymentForms()
+    const billingAddressForm = useForm({
+        mode: 'onChange',
+        shouldUnregister: false,
+        defaultValues: {...selectedBillingAddress}
+    })
 
     const {removePromoCode, ...promoCodeProps} = usePromoCode()
 
-    useEffect(() => {
-        getPaymentMethods()
-    }, [])
+    const paymentMethodForm = useForm()
+
+    const onPaymentSubmit = async (formValue) => {
+        // The form gives us the expiration date as `MM/YY` - so we need to split it into
+        // month and year to submit them as individual fields.
+        const [expirationMonth, expirationYear] = formValue.expiry.split('/')
+
+        const paymentInstrument = {
+            paymentMethodId: 'CREDIT_CARD',
+            paymentCard: {
+                holder: formValue.holder,
+                issueNumber: formValue.number.replace(/ /g, ''),
+                cardType: getPaymentInstrumentCardType(formValue.cardType),
+                expirationMonth: parseInt(expirationMonth),
+                expirationYear: parseInt(`20${expirationYear}`)
+            }
+        }
+
+        return addPaymentInstrumentToBasket({
+            parameters: {basketId: basket.basketId},
```

Comment:
```suggestion
            parameters: {basketId: basket?.basketId},
```
And there are other similar cases in this file too.